[RFC] activation checkpointing: enable mixing tensor without requires_grad

zhengwy888 opened the pull request:
Summary:
https://pytorch.org/docs/stable/checkpoint.html pytorch checkpoint requires all input to the function being checkpointed to requires_grad, but this assumption is not necessarily try. consider the following two examples

```
output = MultiheadedMaskedAtten(input, mask)

output = LSTM(input, seq_length)
```
both length and mask are tensors that won't requires grad, currently if you try to checkpoint torch.autograd.backward will complain

```
  File "/mnt/xarfuse/uid-124297/7d159c34-seed-nspid4026531836-ns-4026531840/torch/autograd/function.py
", line 87, in apply
    return self._forward_cls.backward(self, *args)
  File "/mnt/xarfuse/uid-124297/7d159c34-seed-nspid4026531836-ns-4026531840/torch/utils/checkpoint.py"
, line 99, in backward
    torch.autograd.backward(outputs, args)
  File "/mnt/xarfuse/uid-124297/7d159c34-seed-nspid4026531836-ns-4026531840/torch/autograd/__init__.py
", line 132, in backward
    allow_unreachable=True)  # allow_unreachable flag
RuntimeError: element 1 of tensors does not require grad and does not have a grad_fn
```

In theory we can just skip the non-grad-requiring tensor when running autograd.backward. this diff allows this skipping, but I don't know have insights into what other problem this might cause, so looking for pytorch reviewers for comments.

Test Plan:
locally tested my checkpoint model (in the next diff)
unit test TODO after gathering comments.

Differential Revision: D24094764

albanD commented on torch/utils/checkpoint.py:
```diff
@@ -96,7 +96,14 @@ def backward(ctx, *args):
+            if outputs[i].requires_grad:
+                outputs_with_grad.append(outputs[i])
+                args_with_grad.append(args[i])
+        torch.autograd.backward(outputs_with_grad, args_with_grad)
```
What if `len(outputs_with_grad) == 0`?

zhengwy888 replied:
conceptually if none of the output requires grad then checkpoint shouldn't be used in the first place. I will raise an exception if all output doesn't require grad.

zhengwy888 replied:
the current behavior is throwing an warning in check_backward_validity(), though in fact backward pass would actually fail without my change. 
should I keep the warning or change that to an exception?

albanD reviewed (commented):
This sounds like an acceptable new feature for checkpoint.

You will need to add a few things:
- Add extra checks when nothing requires_grad or other edge cases. Make sure that things that used to be invalid are still invalid except what we document.
- Document the new changes
- Test the new changes

zhengwy888 commented:
thanks for the review @albanD. This is my first pytorch diff, could you provide a bit more detail regarding 'Document the new changes'? would some comments around the addition be enough?

albanD reviewed (commented):
Hi,

I think you can add a paragraph in the doc [here](https://github.com/pytorch/pytorch/blob/11334e1df89480eab4b9b33690b80fe73c8a05dc/torch/utils/checkpoint.py#L127) explaining that all outputs will require gradients after the checkpoint in the forward but only the outputs truly requiring gradients will actually be have gradients computed for them.

And you can add tests in the `test_autograd.py` next to the other checkpoint() tests there to ensure that the new errors are properly raised and that what you want is working fine.

zhengwy888 commented:
I pushed an updated commit at https://github.com/zhengwy888/pytorch/commit/50de2c14c3e566d7c50459263332123546b65423 but it seems it didn't merge with this pull request. do I have to do something?

albanD commented:
Hi,

You should just update your local branch and then push the PR.
If this is exported from fbcode, you might have to update the diff first then re-export? (I have never done that)

zhengwy888 commented:
looks like there was a delay, now this pull request has been updated, and ready for review.

albanD commented on test/test_utils.py:
```diff
@@ -268,6 +268,20 @@ def run_fn(tensor1, tensor2):
+        def run_fn(tensor1, tensor2):
+            # tensor 2 is used for other application logic
+            return tensor1
+        input_var = torch.randn(1, 100, requires_grad=True)
```
nit: please use a smaller size here as there is no need for such a big Tensor. Size of 2 or 4 would be enough.

albanD commented on test/test_utils.py:
```diff
@@ -268,6 +268,20 @@ def run_fn(tensor1, tensor2):
+            return tensor1
+        input_var = torch.randn(1, 100, requires_grad=True)
+        input_var2 = torch.randn(1, 100, requires_grad=False)
+        out = checkpoint(run_fn, input_var, input_var2)
```
This would work with the old version of the code no? The old code only fails if an output of the function does not require gradients right?

albanD commented on torch/utils/checkpoint.py:
```diff
@@ -96,7 +96,16 @@ def backward(ctx, *args):
+            if outputs[i].requires_grad:
+                outputs_with_grad.append(outputs[i])
+                args_with_grad.append(args[i])
+        assert len(outputs_with_grad) > 0, "none of output has requires_grad=True, this checkpoint() is not necessary"
```
nit: we usually throw RuntimeError for these kind of things.

Commit:
[RFC] activation checkpointing: enable mixing tensor without requires_grad (#45934)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/45934

https://pytorch.org/docs/stable/checkpoint.html pytorch checkpoint requires all input to the function being checkpointed to requires_grad, but this assumption is not necessarily try. consider the following two examples

```
output = MultiheadedMaskedAtten(input, mask)

output = LSTM(input, seq_length)
```
both length and mask are tensors that won't requires grad, currently if you try to checkpoint torch.autograd.backward will complain

```
  File "/mnt/xarfuse/uid-124297/7d159c34-seed-nspid4026531836-ns-4026531840/torch/autograd/function.py
", line 87, in apply
    return self._forward_cls.backward(self, *args)
  File "/mnt/xarfuse/uid-124297/7d159c34-seed-nspid4026531836-ns-4026531840/torch/utils/checkpoint.py"
, line 99, in backward
    torch.autograd.backward(outputs, args)
  File "/mnt/xarfuse/uid-124297/7d159c34-seed-nspid4026531836-ns-4026531840/torch/autograd/__init__.py
", line 132, in backward
    allow_unreachable=True)  # allow_unreachable flag
RuntimeError: element 1 of tensors does not require grad and does not have a grad_fn
```

this diff allows skipping the non-grad-requiring tensor when running autograd.backward.

added documentation for this feature as well.

Test Plan: added unit test to make sure partial tensor grads can be used in checkpoint().

Differential Revision: D24094764

fbshipit-source-id: f45bf4146ba9753f4945bc2519dd20f06f3949ee

zhengwy888 commented:
this diff is ready for review now.

albanD reviewed (approved):
LGTM thanks for the updates.
